### DISCUSSION TITLE
Cannot use Valeronoi on Roborock S7 because signal strenght is always 0

### DISCUSSION THREAD

**Discussion by Alabate** [Hypfer/Valetudo#1724] [Q&A / Support] [answered]
Hello,

I tried to use Valeronoi to map the wifi at my place with the Roborock S7. But it does nothing because doing GET on `/api/v2/robot/capabilities/WifiConfigurationCapability` always return a signal strength of 0:

```
{"__class":"ValetudoWifiStatus","metaData":{},"state":"connected","details":{"bssid":"xx:xx:xx:xx:xx:xx","ssid":"AlabateWifi","signal":0,"frequency":"2.4ghz","ips":["10.xx.xx.xxx","fe80::xxxx:xxxx:xxxx:xxxx"]}}
```

Is this value always supposed to be available? Or maybe the signal strength is just not supported on the Roborock S7?

Anyway even if that's not supported, thank for your work, Valetudo is a really stable and pleasent to use, and this is what matters!

**Comment by Hypfer** [comment, marked as answer]
Unfortunately, that seems to be an issue of the S7 firmware. You're the second user to report that.
The first one ran `iw dev wlan0 link` manually and found that it simply reported 0dBm as the signal strength.

=> Probably driver-related.

**Reply by Alabate** [reply]
I can confirm the output of `iw dev wlan0 link`:
```
[root@rockrobo ~]# iw dev wlan0 link
Connected to xx:xx:xx:xx:xx:xx (on wlan0)
	SSID: AlabateWifi
	freq: 2412
	RX: 0 bytes (0 packets)
	TX: 0 bytes (0 packets)
	signal: 0 dBm
	tx bitrate: 
```
So yeah I agree that this is probably driver related.

----

However, we can get the the signal strength by running `iw dev wlan0 scan`. So I managed to *patch* `iw` output, the tricky part is to make Valetudo use our patched `iw`. Here is what I did:
- On the robot, create the file `/mnt/data/patched-bin/iw`:
```
#!/bin/bash
###############################################################################
# Patch 'iw' output on the fly to get a valid signal strength on Roborock S7
###############################################################################
# The signal strength is not available on this device when running
#     iw dev [DEVICE] link
# but it is available when doing a wifi scan.
# To make the scan as fast as possible we can scan only the current frequency:
#     iw dev [DEVICE] scan freq [FREQUENCY]
# We can then patch the output of iw to make the signal strength available when
# Valetudo run the 'link' command.
#
# See https://github.com/Hypfer/Valetudo/discussions/1724

# Only patch the output for "iw dev [DEVICE] link"
if [[ "$#" == 3 ]] && [[ "$1" == "dev" ]] && [[ "$3" == "link" ]] ; then
	dev="$2"
	freq=$(/sbin/iw dev $dev link | grep 'freq: ' | awk '{print $NF}')
	SSID=$(/sbin/iw dev $dev link | grep 'SSID: ' | awk '{print $NF}')
	signal_strength=$(/sbin/iw dev $dev scan freq $freq | grep -E "SSID: $SSID\$" -B2 | grep -oE 'signal: [-0-9\.]+ ' | awk '{print $NF}')
	out_str="$(/sbin/iw dev $dev link)"
	exit_code="$?"
	
	echo "$out_str" | sed "s/signal: 0 dBm/\signal: $signal_strength dBm/"
	exit $exit_code
else
	exec /sbin/iw "$@"
fi
```
- Make it executable with `chmod +x /mnt/data/patched-bin/iw`
- Then edit (with `vi`) `/mnt/reserve/_root.sh` and add the following line, just before `VALETUDO_CONFIG_PATH`
```
PATH="/mnt/data/patched-bin:$PATH"
```
- Reboot your robot and you should now have a non-zero signal value and Valeronoi should work.
----
I don't know if a hack like this could be integrated into the project, and if it's the case it will be probably easier and safer to integrate it directly into Valetudo. But I don't know if you accept device-specific hack like this. If yes, I could probably make a PR for that if you want.

**Reply by Hypfer** [reply]
Given that no one noticed previously, it might be a bug in the 1612 firmware and thus there's hope that it will go away in the future.
I wouldn't spend too much effort on that

Note that that script only exclusively seems to work with the S7 which for some reason seems to have scan results available instantly. Other robots might need up to a whole minute for the `iw scan` to return. Sometimes, it just doesn't return at all due to driver bugs. Thus, this script would actually require a few more checks and timeouts if the S7 ever decides to not behave like it currently seems to do.

Really not worth the trouble. What could be worth considering though is extending Valeronoi to usw the WiFiScanCapability. I think that was also the plan at some point

The code for that capability is also already resilient against broken Wi-Fi drivers and on top of that, you can map out any network the robot can see instead of just the one it is connected to.
https://github.com/Hypfer/Valetudo/blob/master/backend/lib/robots/common/linuxCapabilities/LinuxWifiScanCapability.js
